src/_pytest/python_api.py

```diff
@@ -273,8 +273,20 @@ def __eq__(self, actual) -> bool:
         if math.isinf(abs(self.expected)):  # type: ignore[arg-type]
             return False
 
-        # Return true if the two numbers are within the tolerance.
-        result: bool = abs(self.expected - actual) <= self.tolerance
+        # Return true if the two numbers are within the tolerance.  In order to
+        # be flexible about which types are supported, try making the
+        # comparison in two ways.  The first requires that the actual value can
+        # be subtracted from the expected value.  This is necessary for complex
+        # numbers, which don't implement comparison operators.  The second
+        # requires that the actual and expected values can be compared.  This
+        # is necessary for comparing Decimals with floats, see #8495.
+        try:
+            result: bool = abs(self.expected - actual) <= self.tolerance
+        except TypeError:
+            low = self.expected - self.tolerance
+            high = self.expected + self.tolerance
+            result: bool = low <= actual <= high
```

Comment:
if we encounter a `TypeError` here we should not give users a false confidence that this works (when it only works sometimes)

I've demonstrated a specific case here: https://github.com/pytest-dev/pytest/issues/8495#issuecomment-812688829